cellprofiler/modules/plugins/imagetemplate.py

```diff
@@ -176,128 +276,110 @@ def create_settings(self):
     # pipeline file.
     #
     def settings(self):
-        return [self.input_image_name, self.output_image_name,
-                self.gradient_choice, self.automatic_smoothing,
-                self.scale]
+        #
+        # The superclass's "settings" method returns [self.x_name, self.y_name],
+        # which are the input and output image settings.
+        #
+        settings = super(ImageTemplate, self).settings()
+
+        # Append additional settings here.
+        return settings + [
+            self.gradient_choice,
+            self.automatic_smoothing,
+            self.scale
+        ]
 
     #
-    # visible_settings tells CellProfiler which settings should be
+    # "visible_settings" tells CellProfiler which settings should be
     # displayed and in what order.
     #
     # You don't have to implement "visible_settings" - if you delete
     # visible_settings, CellProfiler will use "settings" to pick settings
     # for display.
     #
     def visible_settings(self):
-        result = [self.input_image_name, self.output_image_name,
-                  self.gradient_choice, self.automatic_smoothing]
+        #
+        # The superclass's "visible_settings" method returns [self.x_name,
+        # self.y_name], which are the input and output image settings.
+        #
+        visible_settings = super(ImageTemplate, self).visible_settings()
+
+        # Configure the visibility of additional settings below.
+        visible_settings += [
+            self.gradient_choice,
+            self.automatic_smoothing
+        ]
+
         #
         # Show the user the scale only if self.wants_smoothing is checked
         #
         if not self.automatic_smoothing:
-            result += [self.scale]
-        return result
+            visible_settings += [self.scale]
+
+        return visible_settings
 
     #
     # CellProfiler calls "run" on each image set in your pipeline.
-    # This is where you do the real work.
     #
     def run(self, workspace):
         #
-        # Get the input and output image names. You need to get the .value
-        # because otherwise you'll get the setting object instead of
-        # the string name.
-        #
-        input_image_name = self.input_image_name.value
-        output_image_name = self.output_image_name.value
-        #
-        # Get the image set. The image set has all of the images in it.
-        #
-        image_set = workspace.image_set
-        #
-        # Get the input image object. We want a grayscale image here.
-        # The image set will convert a color image to a grayscale one
-        # and warn the user.
+        # The superclass's "run" method handles retreiving the input image
+        # and saving the output image. Module-specific behavior is defined
+        # by setting "self.function", defined in this module. "self.function"
+        # is called after retrieving the input image and before saving
+        # the output image.
```

Comment:
I know that in some cases one needs access to other parts of `self` that aren't available in `self.function`. Do you think it would be beneficial to add that here or is this template supposed to be a fairly simple template?

Reply:
Good question. This is intended to be a very simple template. Accessing other methods or instance variables is beyond the scope of this example.

I'm working on a guide for writing modules. I expect to include information on writing a module with a function that needs state (`self`).